### PR TITLE
✨ Added Support for actions after flow is completed

### PR DESCRIPTION
**Pull request by andretortolano**
* I added actionProcessor as the StateMachine dependency.
* Added Tests to cover this new use case of the statemachine

### CODE REVIEW

File: appcues/src/main/java/com/appcues/action/ActionProcessor.kt [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 import com.appcues.AppcuesCoroutineScope
 import kotlinx.coroutines.channels.Channel
 import kotlinx.coroutines.launch
+import org.koin.core.scope.Scope
 
 // Responsible for queueing up any ExperienceActions to run in sequential order,
 // to allow for things like closing current experience + launching a new experience, for example.
 internal class ActionProcessor(
-    private val appcues: Appcues,
+    private val scope: Scope,
```
Comment (andretortolano) [inline appcues/src/main/java/com/appcues/action/ActionProcessor.kt:11]: had to change this so we avoid circular dependency.

File: appcues/src/main/java/com/appcues/analytics/ExperienceLifecycleTracker.kt [hunk 2]
```diff
@@ -57,12 +58,14 @@ internal class ExperienceLifecycleTracker(
                         trackLifecycleEvent(StepCompleted(it.experience, it.flatStepIndex))
                     }
                     is EndingExperience -> {
-                        if (it.markComplete || it.flatStepIndex == it.experience.flatSteps.count() - 1) {
+                        if (it.isExperienceCompleted()) {
                             // if ending on the last step OR an action requested it be considered complete explicitly,
                             // track the experience_completed event
+                            Log.i("Appcues", "Experience ${it.experience.name} Completed")
```
Comment (andretortolano) [inline appcues/src/main/java/com/appcues/analytics/ExperienceLifecycleTracker.kt:64]: logs will be removed.

File: appcues/src/main/java/com/appcues/data/mapper/experience/ExperienceMapper.kt [hunk 3]
```diff
@@ -40,6 +43,10 @@ internal class ExperienceMapper(
             priority = priority,
             type = from.type,
             publishedAt = from.publishedAt,
+            completionActions = arrayListOf<ExperienceAction>().apply {
```
Comment (andretortolano) [inline appcues/src/main/java/com/appcues/data/mapper/experience/ExperienceMapper.kt:46]: this is where I'm mapping from the remote data and creating the ExperienceActions as a list. (the order here matters)

File: appcues/src/main/java/com/appcues/statemachine/State.kt [hunk 4]
```diff
@@ -24,6 +25,11 @@ internal sealed class State {
         // and it should be set to null if a dismiss is not requested (i.e. moving to next step in same container)
         val dismissAndContinue: (() -> Unit)?,
     ) : State()
-    data class EndingExperience(val experience: Experience, val flatStepIndex: Int, val markComplete: Boolean) : State()
+
+    data class EndingExperience(val experience: Experience, val flatStepIndex: Int, val markComplete: Boolean) : State() {
+        // this defines whether the experience was completed or dismissed
+        fun isExperienceCompleted() = markComplete || flatStepIndex == experience.flatSteps.count() - 1
```
Comment (andretortolano) [inline appcues/src/main/java/com/appcues/statemachine/State.kt:31]: moved this logic to the data classe so it can be reused in two places.

File: appcues/src/main/java/com/appcues/statemachine/Transitions.kt [hunk 5]
```diff
@@ -144,7 +148,7 @@ internal interface Transitions {
     }
 
     fun EndingExperience.fromEndingExperienceToIdling(action: Reset): Transition {
-        return Transition(Idling)
+        return Transition(Idling, if (isExperienceCompleted()) ProcessActions(experience.completionActions) else null)
```
Comment (andretortolano) [inline appcues/src/main/java/com/appcues/statemachine/Transitions.kt:151]: if experience is completed we ProcessActions or else we do nothing.

File: appcues/src/main/java/com/appcues/action/ActionKoin.kt [hunk 6]
```diff
@@ -14,7 +14,7 @@ internal object ActionKoin : KoinScopePlugin {
 
     override fun ScopeDSL.install() {
         scoped { ActionRegistry(scope = get(), logcues = get()) }
-        scoped { ActionProcessor(appcues = get(), appcuesCoroutineScope = get()) }
+        scoped { ActionProcessor(scope = get(), appcuesCoroutineScope = get()) }
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/action/ActionKoin.kt:17]: this one (and the line above) - if we we are passing the scope in to avoid circular dependencies - maybe we should just not pass any other koin scope dependencies in any longer? Then resolve them as injected props with scope.inject()? It feels strange to have a mix in the same class.

File: appcues/src/main/java/com/appcues/action/appcues/LaunchExperienceAction.kt [hunk 7]
```diff
@@ -9,7 +9,14 @@ internal class LaunchExperienceAction(
     override val config: AppcuesConfigMap,
 ) : ExperienceAction {
 
+    constructor(experienceId: String) : this(
+        hashMapOf<String, Any>().apply {
+            this["experienceID"] = experienceId
+        }
+    )
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/action/appcues/LaunchExperienceAction.kt:16]: nice way to tuck this away 👍

File: appcues/src/main/java/com/appcues/action/ActionKoin.kt (hunk 6 above)
Comment (andretortolano) [reply appcues/src/main/java/com/appcues/action/ActionKoin.kt:17]: will do thanks

File: appcues/src/main/java/com/appcues/action/ActionKoin.kt [hunk 8]
```diff
@@ -14,7 +14,7 @@ internal object ActionKoin : KoinScopePlugin {
 
     override fun ScopeDSL.install() {
         scoped { ActionRegistry(scope = get(), logcues = get()) }
-        scoped { ActionProcessor(appcues = get(), appcuesCoroutineScope = get()) }
+        scoped { ActionProcessor(scope = get()) }
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/action/ActionKoin.kt:17]: not part of this PR, but the `logcues` thing on the ActionRegistry line above is similar - mix of injection patterns

File: appcues/src/main/java/com/appcues/action/appcues/LinkAction.kt [hunk 9]
```diff
@@ -7,16 +7,25 @@ import com.appcues.data.model.AppcuesConfigMap
...
 
     companion object {
+
         const val TYPE = "@appcues/link"
     }
 
+    constructor(redirectUrl: String, linkOpener: LinkOpener) : this(
+        hashMapOf<String, Any>().apply {
+            this["url"] = redirectUrl
+        },
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/action/appcues/LinkAction.kt:25]: nit - is the apply needed here? 
```
hashMapOf<String, Any>("url" to redirectUrl)
```

File: appcues/src/main/java/com/appcues/analytics/ExperienceLifecycleTracker.kt [hunk 10]
```diff
@@ -57,7 +57,7 @@ internal class ExperienceLifecycleTracker(
                         trackLifecycleEvent(StepCompleted(it.experience, it.flatStepIndex))
                     }
                     is EndingExperience -> {
-                        if (it.markComplete || it.flatStepIndex == it.experience.flatSteps.count() - 1) {
+                        if (it.isExperienceCompleted()) {
```
Comment (iujames) [inline appcues/src/main/java/com/appcues/analytics/ExperienceLifecycleTracker.kt:60]: oh good catch re-using this

**Review by iujames — APPROVED** [review]

File: appcues/src/main/java/com/appcues/action/appcues/LinkAction.kt (hunk 9 above)
Comment (andretortolano) [reply appcues/src/main/java/com/appcues/action/appcues/LinkAction.kt:25]: oh you are right

File: appcues/src/main/java/com/appcues/action/ActionKoin.kt (hunk 8 above)
Comment (andretortolano) [reply appcues/src/main/java/com/appcues/action/ActionKoin.kt:17]: I will fix this as well
